File: Android/samples/sdksample/app/src/main/java/com/microsoft/maps/sample/LocalSearch.java
package com.microsoft.maps.sample;

import android.content.Context;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import com.microsoft.maps.GeoboundingBox;
import com.microsoft.maps.Geolocation;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

class LocalSearch {

    private static final String URL_ENDPOINT
            = "https://dev.virtualearth.net/REST/v1/LocalSearch/"
            + "?query={query}"
            + "&userMapView={bounds}"
            + "&key=" + BuildConfig.CREDENTIALS_KEY;

    static class Poi {
        String name;
        Geolocation location;
    }

    interface Callback {
        void onSuccess(List<Poi> results);
        void onFailure();
    }

    static void sendRequest(Context context, String query, GeoboundingBox bounds, Callback callback) {
        if (query == null || query.isEmpty()) {
            Toast.makeText(context, "Invalid query", Toast.LENGTH_LONG).show();
            return;
        }

        String boundsStr = String.format(Locale.ROOT,
            "%.6f,%.6f,%.6f,%.6f",
            bounds.south, bounds.west, bounds.north, bounds.east);

        RequestQueue queue = Volley.newRequestQueue(context);
        queue.add(new StringRequest(
            Request.Method.GET,
            URL_ENDPOINT.replace("{query}", query).replace("{bounds}", boundsStr),
            (String response) -> {
                List<Poi> results = parse(response);
                if (results == null || results.isEmpty()) {
                    callback.onFailure();
                    return;
                }
                callback.onSuccess(results);
            },
            (VolleyError error) -> callback.onFailure()
        ));
    }

    private static List<Poi> parse(String json) {
        List<Poi> results = new ArrayList<>();

        try {
            JSONObject responseJObj = new JSONObject(json);
            JSONArray resourceSetsJArr = responseJObj.getJSONArray("resourceSets");
            JSONObject resourceSetJObj = resourceSetsJArr.getJSONObject(0);
            JSONArray resourcesJArr = resourceSetJObj.getJSONArray("resources");

            for (int i = 0; i < resourcesJArr.length(); ++i) {
                JSONObject resourceJObj = resourcesJArr.getJSONObject(i);
                Poi poi = new Poi();

                poi.name = resourceJObj.getString("name");

                JSONObject pointJObj = resourceJObj.getJSONObject("point");
                JSONArray coordinatesJArr = pointJObj.getJSONArray("coordinates");
                poi.location = new Geolocation(
                        coordinatesJArr.getDouble(0),
                        coordinatesJArr.getDouble(1));

                results.add(poi);
            }
        }
        catch (Exception ex) {
            return null;
        }

        return results;
    }

}

File: iOS/samples/sdksample/BingMapsSdkSample/LocalSearch.swift
import UIKit
import MicrosoftMaps

class LocalSearch: NSObject {
    static let CREDENTIALS_KEY = Bundle.main.infoDictionary?["CREDENTIALS_KEY"] as! String
    static let URL_ENDPOINT = "https://dev.virtualearth.net/REST/v1/LocalSearch/"
    + "?query={query}"
    + "&userMapView={south},{west},{north},{east}"
    + "&key=" + CREDENTIALS_KEY

    struct Poi {
        var name: NSString!
        var location:MSGeopoint!
    }

    static func sendRequest(queryOptional:String?, bounds:MSGeoboundingBox, _completion: @escaping (Array<Poi>?) -> Void ) {
        guard let query = queryOptional else {
            _completion(nil)
            return
        }
        if (query == "") {
            _completion(nil)
            return
        }
        var urlString = URL_ENDPOINT.replacingOccurrences(of: "{query}", with: query)
        urlString = urlString.replacingOccurrences(of: "{south}", with: String(bounds.south))
        urlString = urlString.replacingOccurrences(of: "{west}", with: String(bounds.west))
        urlString = urlString.replacingOccurrences(of: "{north}", with: String(bounds.north))
        urlString = urlString.replacingOccurrences(of: "{east}", with: String(bounds.east))

        guard let url = URL(string: urlString) else {
            _completion(nil)
            return
        }

        let request = URLRequest(url: url)
        let session = URLSession.shared

        session.dataTask(with: request, completionHandler: { data, response, error -> Void in
            DispatchQueue.main.async {
                _completion(parse(data:data))
            }
        }).resume()
    }

    static func parse(data:Data?) -> (Array<Poi>) {
        var results = Array<Poi>()

        do {
            let responseJObj = try JSONSerialization.jsonObject(with:data!, options:[]) as? NSDictionary
            let resourceSetsJArr = responseJObj!["resourceSets"] as! NSArray

            if resourceSetsJArr.count > 0 {
                let resourceSetJObj = resourceSetsJArr[0] as? NSDictionary
                let resourcesJArr = resourceSetJObj!["resources"] as! NSArray

                for resourceObj in resourcesJArr {
                    let resourceObjDictionary = resourceObj as! NSDictionary
                    // some results do not contain point
                    if resourceObjDictionary["point"] != nil {
                        var poi = Poi()
                        poi.name = resourceObjDictionary["name"] as? NSString
                        let pointJObj = resourceObjDictionary["point"] as! NSDictionary
                        let coordinatesJArr = pointJObj["coordinates"] as! NSArray
                        poi.location = MSGeopoint(latitude:coordinatesJArr[0] as! Double, longitude:coordinatesJArr[1] as! Double, altitude: 0, altitudeReferenceSystem: .surface)
                        results.append(poi)
                    }
                }
            }
        }
        catch {
        }

        return results;
    }
}
